File: infra-market-quarkus/src/main/kotlin/io/infra/qk/service/RoleService.kt
package io.infra.qk.service

import io.infra.qk.dto.ApiResponse
import io.infra.qk.dto.RoleFormDto
import io.infra.qk.dto.RoleQueryDto
import io.infra.qk.dto.PageResultDto
import io.infra.qk.dto.RoleDto
import io.infra.qk.entity.Role
import io.infra.qk.entity.RolePermission
import io.infra.qk.enums.StatusEnum
import io.infra.qk.repository.RoleRepository
import io.infra.qk.repository.RolePermissionRepository
import io.infra.qk.util.DateTimeUtil
import jakarta.enterprise.context.ApplicationScoped
import jakarta.inject.Inject
import jakarta.transaction.Transactional
import java.util.*

/**
 * 角色服务
 * @author liuqinglin
 * Date: 2025/8/30
 */
@ApplicationScoped
class RoleService @Inject constructor(
    private val roleRepository: RoleRepository,
    private val rolePermissionRepository: RolePermissionRepository
) {
    
    fun getRoleList(query: RoleQueryDto): ApiResponse<PageResultDto<RoleDto>> {
        val (roles, total) = roleRepository.findByCondition(
            query.roleName, query.roleCode, query.status, query.current, query.size
        )
        
        val roleDtos = roles.map { role ->
            RoleDto(
                id = role.id ?: 0,
                roleName = role.name ?: "",
                roleCode = role.code ?: "",
                description = role.description,
                status = role.status,
                createTime = DateTimeUtil.formatDateTime(role.createTime),
                updateTime = DateTimeUtil.formatDateTime(role.updateTime)
            )
        }
        
        val pageResult = PageResultDto(
            records = roleDtos,
            total = total,
            current = query.current,
            size = query.size
        )
        
        return ApiResponse.success(pageResult)
    }
    
    fun getAllRoles(): ApiResponse<List<RoleDto>> {
        val roles = roleRepository.findAllActive()
        
        val roleDtos = roles.map { role ->
            RoleDto(
                id = role.id ?: 0,
                roleName = role.name ?: "",
                roleCode = role.code ?: "",
                description = role.description,
                status = role.status,
                createTime = DateTimeUtil.formatDateTime(role.createTime),
                updateTime = DateTimeUtil.formatDateTime(role.updateTime)
            )
        }
        
        return ApiResponse.success(roleDtos)
    }
    
    fun getRoleById(id: Long): ApiResponse<RoleDto> {
        val role = roleRepository.findById(id) ?: return ApiResponse.error("角色不存在")
        
        val roleDto = RoleDto(
            id = role.id ?: 0,
            roleName = role.name ?: "",
            roleCode = role.code ?: "",
            description = role.description,
            status = role.status,
            createTime = DateTimeUtil.formatDateTime(role.createTime),
            updateTime = DateTimeUtil.formatDateTime(role.updateTime)
        )
        
        return ApiResponse.success(roleDto)
    }
    
    @Transactional
    fun createRole(roleForm: RoleFormDto): ApiResponse<Unit> {
        // 检查角色编码是否已存在
        if (roleRepository.existsByRoleCode(roleForm.roleCode)) {
            return ApiResponse.error("角色编码已存在")
        }
        
        val role = Role(
            name = roleForm.roleName,
            code = roleForm.roleCode,
            description = roleForm.description,
            status = roleForm.status
        )
        
        roleRepository.persist(role)
        
        // 保存角色权限关联
        saveRolePermissions(role.id ?: 0, roleForm.permissionIds)
        
        return ApiResponse.success()
    }
    
    @Transactional
    fun updateRole(id: Long, roleForm: RoleFormDto): ApiResponse<Unit> {
        val role = roleRepository.findById(id) ?: return ApiResponse.error("角色不存在")
        
        // 检查角色编码是否已被其他角色使用
        val existingRole = roleRepository.findByRoleCode(roleForm.roleCode)
        if (existingRole != null && existingRole.id != id) {
            return ApiResponse.error("角色编码已存在")
        }
        
        role.name = roleForm.roleName
        role.code = roleForm.roleCode
        role.description = roleForm.description
        role.status = roleForm.status
        role.updateTime = Date()
        
        roleRepository.persist(role)
        
        // 更新角色权限关联
        rolePermissionRepository.deleteByRoleId(id)
        saveRolePermissions(id, roleForm.permissionIds)
        
        return ApiResponse.success()
    }
    
    @Transactional
    fun deleteRole(id: Long): ApiResponse<Unit> {
        val role = roleRepository.findById(id) ?: return ApiResponse.error("角色不存在")
        
        // 删除角色权限关联
        rolePermissionRepository.deleteByRoleId(id)
        
        // 删除角色
        roleRepository.deleteById(id)
        
        return ApiResponse.success()
    }
    
    @Transactional
    fun updateRoleStatus(id: Long, status: String): ApiResponse<Unit> {
        val role = roleRepository.findById(id) ?: return ApiResponse.error("角色不存在")
        
        role.status = status
        role.updateTime = Date()
        roleRepository.persist(role)
        
        return ApiResponse.success()
    }
    
    @Transactional
    fun batchDeleteRoles(ids: List<Long>): ApiResponse<Unit> {
        ids.forEach { id ->
            deleteRole(id)
        }
        return ApiResponse.success()
    }
    
    private fun saveRolePermissions(roleId: Long, permissionIds: List<Long>) {
        permissionIds.forEach { permissionId ->
            val rolePermission = RolePermission(roleId = roleId, permissionId = permissionId)
            rolePermissionRepository.persist(rolePermission)
        }
    }
}

File: infra-market-server-dotnet/Services/RoleService.cs
using InfraMarket.Server.Data;
using InfraMarket.Server.DTOs;
using InfraMarket.Server.Entities;
using InfraMarket.Server.Repositories.Interfaces;
using InfraMarket.Server.Services.Interfaces;
using InfraMarket.Server.Utils;

namespace InfraMarket.Server.Services;

public class RoleService(
    ApplicationDbContext context,
    IRoleRepository roleRepository,
    IRolePermissionRepository rolePermissionRepository,
    IUserRoleRepository userRoleRepository) : IRoleService
{
    public async Task<ApiData<PageResult<RoleDto>>> GetRolesAsync(RoleQueryDto query)
    {
        var (roles, total) = await roleRepository.PageAsync(query);
        var roleIds = roles.Select(r => r.Id).ToList();
        var rolePermissions = await rolePermissionRepository.FindByRoleIDsAsync(roleIds);
        var rolePermissionMap = rolePermissions.GroupBy(rp => rp.RoleId)
            .ToDictionary(g => g.Key, g => g.Select(rp => rp.PermissionId).ToList());

        var roleDtos = roles.Select(role =>
        {
            var permissionIds = rolePermissionMap.GetValueOrDefault(role.Id, []);
            return ConvertRoleToDto(role, permissionIds);
        }).ToList();

        var result = new PageResult<RoleDto>
        {
            Records = roleDtos,
            Total = total,
            Page = query.Page ?? 1,
            Size = query.Size ?? 10
        };

        return ApiData<PageResult<RoleDto>>.Success(result);
    }

    public async Task<ApiData<List<RoleDto>>> GetAllRolesAsync()
    {
        var roles = await roleRepository.FindByStatusAsync("active");
        var roleIds = roles.Select(r => r.Id).ToList();
        var rolePermissions = await rolePermissionRepository.FindByRoleIDsAsync(roleIds);
        var rolePermissionMap = rolePermissions.GroupBy(rp => rp.RoleId)
            .ToDictionary(g => g.Key, g => g.Select(rp => rp.PermissionId).ToList());

        var roleDtos = roles.Select(role =>
        {
            var permissionIds = rolePermissionMap.GetValueOrDefault(role.Id, []);
            return ConvertRoleToDto(role, permissionIds);
        }).ToList();

        return ApiData<List<RoleDto>>.Success(roleDtos);
    }

    public async Task<ApiData<RoleDto>> GetRoleAsync(ulong id)
    {
        var role = await roleRepository.FindByIdAsync(id);
        if (role == null)
        {
            return ApiData<RoleDto>.Error("角色不存在", 404);
        }

        var rolePermissions = await rolePermissionRepository.FindByRoleIdAsync(id);
        var permissionIds = rolePermissions.Select(rp => rp.PermissionId).ToList();
        var roleDto = ConvertRoleToDto(role, permissionIds);

        return ApiData<RoleDto>.Success(roleDto);
    }

    public async Task<ApiData<RoleDto>> CreateRoleAsync(RoleFormDto dto)
    {
        var existingRole = await roleRepository.FindByCodeAsync(dto.Code);
        if (existingRole != null)
        {
            return ApiData<RoleDto>.Error("角色编码已存在", 400);
        }

        var role = new Role
        {
            Name = dto.Name,
            Code = dto.Code,
            Description = dto.Description,
            Status = "active"
        };

        await using var transaction = await context.Database.BeginTransactionAsync();
        try
        {
            await roleRepository.CreateAsync(role);

            foreach (var permissionId in dto.PermissionIds)
            {
                await rolePermissionRepository.CreateAsync(new RolePermission
                {
                    RoleId = role.Id,
                    PermissionId = permissionId
                });
            }

            await transaction.CommitAsync();
        }
        catch
        {
            await transaction.RollbackAsync();
            throw;
        }

        var roleDto = ConvertRoleToDto(role, dto.PermissionIds);
        return ApiData<RoleDto>.Success(roleDto);
    }

    public async Task<ApiData<RoleDto>> UpdateRoleAsync(ulong id, RoleFormDto dto)
    {
        var role = await roleRepository.FindByIdAsync(id);
        if (role == null)
        {
            return ApiData<RoleDto>.Error("角色不存在", 404);
        }

        var existingRole = await roleRepository.FindByCodeAsync(dto.Code);
        if (existingRole != null && existingRole.Id != role.Id)
        {
            return ApiData<RoleDto>.Error("角色编码已存在", 400);
        }

        role.Name = dto.Name;
        role.Description = dto.Description;

        await using var transaction = await context.Database.BeginTransactionAsync();
        try
        {
            await roleRepository.UpdateAsync(role);
            await rolePermissionRepository.DeleteByRoleIdAsync(id);

            foreach (var permissionId in dto.PermissionIds)
            {
                await rolePermissionRepository.CreateAsync(new RolePermission
                {
                    RoleId = id,
                    PermissionId = permissionId
                });
            }

            await transaction.CommitAsync();
        }
        catch
        {
            await transaction.RollbackAsync();
            throw;
        }

        var roleDto = ConvertRoleToDto(role, dto.PermissionIds);
        return ApiData<RoleDto>.Success(roleDto);
    }

    public async Task<ApiData<object>> DeleteRoleAsync(ulong id)
    {
        var role = await roleRepository.FindByIdAsync(id);
        if (role == null)
        {
            return ApiData<object>.Error("角色不存在", 404);
        }

        if (role.Code == "admin")
        {
            return ApiData<object>.Error("不能删除系统角色", 400);
        }

        var userCount = await userRoleRepository.CountByRoleIdAsync(id);
        if (userCount > 0)
        {
            return ApiData<object>.Error("该角色下还有用户，无法删除", 400);
        }

        role.Status = "deleted";
        await roleRepository.UpdateAsync(role);

        return ApiData<object>.Success(null!);
    }

    public async Task<ApiData<object>> UpdateRoleStatusAsync(ulong id, StatusUpdateDto dto)
    {
        var role = await roleRepository.FindByIdAsync(id);
        if (role == null)
        {
            return ApiData<object>.Error("角色不存在", 404);
        }

        if (role.Code == "admin" && dto.Status == "deleted")
        {
            return ApiData<object>.Error("不能删除系统角色", 400);
        }

        if (role.Status == "deleted" && dto.Status != "deleted")
        {
            return ApiData<object>.Error("已删除的角色不能重新启用", 400);
        }

        if (dto.Status == "deleted")
        {
            return await DeleteRoleAsync(id);
        }

        role.Status = dto.Status;
        await roleRepository.UpdateAsync(role);

        return ApiData<object>.Success(null!);
    }

    public async Task<ApiData<object>> BatchDeleteRolesAsync(BatchRequest request)
    {
        foreach (var id in request.Ids)
        {
            await DeleteRoleAsync(id);
        }

        return ApiData<object>.Success(null!);
    }

    private static RoleDto ConvertRoleToDto(Role role, List<ulong> permissionIds)
    {
        return new RoleDto
        {
            Id = role.Id,
            Name = role.Name,
            Code = role.Code,
            Description = role.Description,
            Status = role.Status,
            PermissionIds = permissionIds,
            CreateTime = DateTimeUtil.FormatLocalDateTime(role.CreateTime),
            UpdateTime = DateTimeUtil.FormatLocalDateTime(role.UpdateTime)
        };
    }
}
